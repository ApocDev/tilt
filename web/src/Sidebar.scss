@import "constants";

.Sidebar {
  position: fixed;
  top: 0;
  right: 0;
  bottom: $statusbar-height;
  width: $sidebar-width;
  background-color: $color-gray;
  border-left: 2px solid $color-gray-lightest;
  box-sizing: border-box;
  overflow-y: auto;
  transform: translateX(0%);
  transition: transform ease $animation-timing;
  font-size: $font-size;
  display: flex;
  flex-direction: column;
  z-index: $z-sidebar;
  box-shadow: -2px 0px 10px 0px rgba($color-gray-dark, $translucent);
}
.Sidebar.is-closed {
  transform: translateX(calc(100% - #{$sidebar-collapsed-width}));
}

// Resource List
.Sidebar-resources {
  flex: 1 0 auto;
  font-weight: bold;
}
.Sidebar-list {
  list-style: none;
}

.resLink {
  right: 0;
  border-bottom: 1px solid $color-gray-light;
  color: $color-white;
  text-decoration: none;
  display: flex;
  align-items: center;
  height: $sidebar-item;
  transition-property: color, background-color;
  transition-duration: $animation-timing;
  transition-timing-function: ease;
}
.resLink:hover {
  right: 0;
  background-color: $color-gray-dark;
  color: $color-blue-light;
}
.resLink.is-selected {
  right: 0;
  background-color: $color-white;
  color: $color-gray;
}
.resLink.has-warnings::after {
  content: "⚠";
  margin-top: -5px;
  margin-left: $spacing-unit / 2;
  font-size: 30px;
  color: $color-yellow;
}
.resLink--all {
  text-transform: uppercase;
}
.resLink--all::before {
  content: "┌";
  color: $color-gray-light;
  margin-top: -8px;
}

.resLink-icon {
  width: $sidebar-item;
  display: flex;
  align-items: center;
  justify-content: center;
}
.resLink--ok .resLink-icon {
  fill: $color-green;
}
.resLink--pending .resLink-icon {
  fill: $color-gray-light;
}
.resLink--error .resLink-icon {
  fill: $color-red;
}
.resLink-name {
  flex: 1 0 auto;
}
.resLink time {
  font-weight: normal;
  color: $color-gray-lightest;
  margin-right: $spacing-unit / 2;
}

// Collapse/Expand
.Sidebar-spacer {
  flex-grow: 1;
  height: 0;
}
.Sidebar-toggle {
  background-color: $color-gray;
  border: 0 none;
  border-top: 1px solid $color-gray-light;
  color: inherit;
  font-size: inherit;
  font-family: inherit;
  display: flex;
  align-items: center;
  min-height: $sidebar-item;
  margin: 0;
  padding: 0;
  @include button-text;
  cursor: pointer;
  transition-property: color, background-color;
  transition-duration: $animation-timing;
  transition-timing-function: ease;
}
.Sidebar-toggle:hover {
  background-color: $color-gray-dark;
  color: $color-blue-light;
}
.Sidebar-toggle > svg {
  fill: $color-white;
  width: $spacing-unit * 1.5;
  transition: transform $animation-timing ease-in, fill $animation-timing ease;
}
.Sidebar-toggle:hover > svg {
  fill: $color-blue-light;
}
.Sidebar.is-closed .Sidebar-toggle > svg {
  transform: rotate(180deg);
}
