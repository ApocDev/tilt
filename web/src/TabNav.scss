@import "constants";

.TabNav {
  position: fixed;
  top: 0;
  left: 0;
  background-color: $color-navy-dark;
  display: flex;
  z-index: $z-tabnav;
}

.viewLink {
  box-sizing: border-box;
  height: $sidebar-item * 1.5;
  flex: 1 0 50%;
  color: inherit;
  text-decoration: none;
  display: flex;
  align-items: center;
  justify-content: center;
  border-bottom: 1px solid $color-navy-light;
  transition-property: color, background-color;
  transition-duration: $animation-timing;
  transition-timing-function: ease;
  @include button-text;
}

.viewLink + .viewLink {
  border-left: 2px solid $color-navy-light;
}

.viewLink:hover {
  background-color: $color-navy-dark;
  color: $color-blue-light;
}

.viewLink--is-selected {
  right: 0;
  background-color: $color-navy;
  border-bottom: none;
}


