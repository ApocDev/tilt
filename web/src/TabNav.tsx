import React, { PureComponent } from "react"
import { Link } from "react-router-dom"
import { SidebarItem } from "./Sidebar"
import { ResourceView } from "./HUD"
import "./TabNav.scss"
type NavProps = {
  resourceName: string
  sidebarItems: Array<SidebarItem>
  resourceView: ResourceView
}

class TabNav extends PureComponent<NavProps> {
  render() {
    let name = this.props.resourceName
    let logResourceViewURL = name === "" ? "/" : `/r/${name}`
    let previewResourceURL = "/"
    if (name) {
      previewResourceURL = `/r/${name}/preview`
    } else if (this.props.sidebarItems.length) {
      // Pick the first item with an endpoint, or default to the first item
      previewResourceURL = `/r/${this.props.sidebarItems[0].name}/preview`
      this.props.sidebarItems.forEach(r => {
        if (r.hasEndpoints) {
          previewResourceURL = `/r/${r.name}/preview`
          return
        }
      })
    }
    let logIsSelected = this.props.resourceView == ResourceView.Log
    let previewIsSelected = this.props.resourceView == ResourceView.Preview
    return (
      <nav className="TabNav">
        <Link
          className={logIsSelected ? "viewLink--is-selected" : ""}
          to={logResourceViewURL}
        >
          Logs
        </Link>
        <Link
          className={previewIsSelected ? "viewLink--is-selected" : ""}
          to={previewResourceURL}
        >
          Preview
        </Link>
      </nav>
    )
  }
}

export default TabNav
