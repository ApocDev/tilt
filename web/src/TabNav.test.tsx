import React from "react"
import renderer from "react-test-renderer"
import TabNav from "./TabNav"
import { ResourceView } from "./HUD"
import { MemoryRouter } from "react-router"
import { oneResourceView, twoResourceView } from "./testdata.test"
import { SidebarItem } from "./Sidebar"

it("doesn't crash with empty resource list", () => {
  let sidebarItems = []
  const tree = renderer
    .create(
      <MemoryRouter>
        <TabNav
          resourceName={"foo"}
          sidebarItems={sidebarItems}
          resourceView={ResourceView.Log}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("previews resources", () => {
  let sidebarItems = []
  const tree = renderer
    .create(
      <MemoryRouter>
        <TabNav
          resourceName={"foo"}
          sidebarItems={sidebarItems}
          resourceView={ResourceView.Preview}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("previews the first resource that has an endpoint if nothing is selected", () => {
  let sidebarItems: Array<SidebarItem> = oneResourceView().Resources.map(
    (res: any) => {
      res.BuildHistory[0].Error = ""
      res.BuildHistory[0].Warnings = ["warning"]
      return new SidebarItem(res)
    }
  )

  const tree = renderer
    .create(
      <MemoryRouter>
        <TabNav
          resourceName={""}
          sidebarItems={sidebarItems}
          resourceView={ResourceView.Preview}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
