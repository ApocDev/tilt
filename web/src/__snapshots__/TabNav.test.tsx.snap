// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't crash with empty resource list 1`] = `
<nav
  className="TabNav"
>
  <a
    className="viewLink--is-selected"
    href="/r/foo"
    onClick={[Function]}
  >
    Logs
  </a>
  <a
    className=""
    href="/r/foo/preview"
    onClick={[Function]}
  >
    Preview
  </a>
</nav>
`;

exports[`previews resources 1`] = `
<nav
  className="TabNav"
>
  <a
    className=""
    href="/r/foo"
    onClick={[Function]}
  >
    Logs
  </a>
  <a
    className="viewLink--is-selected"
    href="/r/foo/preview"
    onClick={[Function]}
  >
    Preview
  </a>
</nav>
`;

exports[`previews the first resource that has an endpoint if nothing is selected 1`] = `
<nav
  className="TabNav"
>
  <a
    className=""
    href="/"
    onClick={[Function]}
  >
    Logs
  </a>
  <a
    className="viewLink--is-selected"
    href="/r/vigoda/preview"
    onClick={[Function]}
  >
    Preview
  </a>
</nav>
`;
